src/main/java/com/allthemods/gravitas2/core/mixin/WorldGenDragonRoostsMixin.java

```diff
@@ -0,0 +1,132 @@
+import com.github.alexthe666.iceandfire.util.WorldUtil;
+import com.github.alexthe666.iceandfire.world.gen.*;
+import com.mojang.serialization.Codec;
+import net.dries007.tfc.common.blocks.TFCBlocks;
+import net.dries007.tfc.common.blocks.rock.Rock;
+import net.dries007.tfc.world.chunkdata.ChunkData;
+import net.dries007.tfc.world.chunkdata.ChunkDataProvider;
+import net.dries007.tfc.world.settings.RockSettings;
+import net.minecraft.core.BlockPos;
+import net.minecraft.util.Mth;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.levelgen.feature.Feature;
+import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
+import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
+import org.jetbrains.annotations.NotNull;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Overwrite;
+import org.spongepowered.asm.mixin.Shadow;
+
+import java.util.Random;
+
+
+@Mixin(value = WorldGenDragonRoosts.class, remap = false)
+public abstract class WorldGenDragonRoostsMixin extends Feature<NoneFeatureConfiguration> implements TypedFeature {
+    private static Block TFCRock = TFCBlocks.ROCK_BLOCKS.get(Rock.BASALT).get(Rock.BlockType.HARDENED).get();
+    private static Block TFCRock2 = TFCBlocks.ROCK_BLOCKS.get(Rock.BASALT).get(Rock.BlockType.HARDENED).get();
+
+    public WorldGenDragonRoostsMixin(Codec<NoneFeatureConfiguration> codec) {
+        super(codec);
+    }
+
+    /**
+     * @author thevortex
+     * @reason
+     */
+    @Overwrite
+    public boolean place(@NotNull FeaturePlaceContext<NoneFeatureConfiguration> context) {
+        EntityType<? extends EntityDragonBase> DRAGONTYPE = this.getDragonType();
+        WorldGenLevel worldIn = context.level();
+        RandomSource rand = context.random();
+        BlockPos pos = context.origin();
+        ChunkDataProvider provider = ChunkDataProvider.get(worldIn);
+        ChunkData data = provider.get(worldIn, pos);
+        float rainfall = data.getRainfall(pos);
+        float avgAnnualTemperature = data.getAverageTemp(pos);
+        RockSettings rocks = data.getRockData().getRock(pos);
+        TFCRock = rocks.hardened();
+        TFCRock2 = rocks.hardened();
+        var climateTest = IAFEntityMap.dragonList.get(DRAGONTYPE);
+        var tempAndRainfall = new float[]{avgAnnualTemperature, rainfall};
+        if (!climateTest.test(tempAndRainfall)) {
+            //GregitasCore.LOGGER.info("Blocked :" + DRAGONTYPE.getDescription() + " at: " + pos);
+            return false;
+        }
+        if (!WorldUtil.canGenerate(IafConfig.generateDragonRoostChance, context.level(), context.random(), context.origin(), this.getId(), true)) {
+            return false;
+        } else {
+            boolean isMale = (new Random()).nextBoolean();
```

Comment:
the featurePlaceContext gives you a RandomSource. use it, so that this is replicatable. if I&F doesn't, skill issue.